### ISSUE TITLE
Thoughts on reading the tutorials

### ISSUE THREAD

**Issue by effigies** [nipype/pydra#119]
These are comments prompted by working through the tutorials, and can be split out into focused issues as needed, but just wanted to start making notes.

The following is from working through `intro_task_state`, and skimming `intro_workflow` enough to write a couple example workflows to demonstrate a point below. I will pick up later.

## Semantic

1. ~~`task1.state.splitter == 'add_two.x'` seems to be using the function name as the `Task` name. Does this cause problems in a workflow where I might use two of the same function-based `Task`s?~~ Realized that tasks in workflows are required to have an explicit name. Wrote this before I'd started on the workflow tutorial.
2. When splitting over multiple dimensions, the results are aggregated in a 1D list. I would have assumed that it would be ND, with the first axis matching the first variable split over, and so on.
3. When combining, the order of the combiner does not affect the order of the results. I would assume that `.combine(["a", "b"])` would produce something ordered like `[(a1, b1), (a2, b1), (a1, b2), (a2, b2)]` while `.combine(["b", "a"])` would produce `[(a1, b1), (a1, b2), (a2, b1), (a2, b2)]`. (Or possibly the reverse; I haven't thought it through carefully enough to be sure.)
4. `.combine("x").combine("n")` seems like it should be equivalent to `.combine(["x", "n"])`. Probably an edge case, but I think making this (and multiple `split`s on a single task) doable should help clarify the semantics for users, as well as our own thinking. For instance, these should be equivalent (they don't actually run, but hopefully they're clear):
```Python
@pydra.mark.task
def power(a, b):
    return a ** b

@pydra.mark.task
def identity(x):
    return x

wf1 = pydra.Workflow(name="wf1", input_spec=["a", "b"], a=[1, 2], b=[2, 3])
wf1.add(power(name="power", a=wf1.lzin.a, b=wf1.lzin.b).split(["a", "b"]).combine("a"))
wf1.add(identity(name="identity", x=wf1.power.lzout.out).combine("power.b"))
wf1.set_output({'out': wf1.identity.lzout.out})

wf2 = pydra.Workflow(name="wf2", input_spec=["a", "b"], a=[1, 2], b=[2, 3])
wf2.add(power(name="power", a=wf2.lzin.a, b=wf2.lzin.b).split(["a", "b"]).combine("a").combine("b"))
wf2.set_output({'out': wf2.power.lzout.out})

wf3 = pydra.Workflow(name="wf3", input_spec=["a", "b"], a=[1, 2], b=[2, 3])
wf3.add(power(name="power", a=wf3.lzin.a, b=wf3.lzin.b).split("a").split("b").combine(["a", "b"])
wf3.set_output({'out': wf3.power.lzout.out})
```

## Usability

1. `State.__str__` would be nice to show something that actually describes the internal state, rather than just `repr(State)`.
2. Combiners must be strings or lists, currently. It's not clear why this is a requirement. Any sequence seems acceptable to me.
3. `nr_proc` is pretty unintuitive.

## Aesthetic

1. Perhaps we can think of a more intuitive name from a user's perspective than `splitter` in `Task().split(splitter=...)`. I think it's fine as an internal variable, so we don't need to do a massive search-and-replace, but `splitter` kind of grates on me.
   * Although parameters are usually given noun names, we could follow Pandas use a preposition like `split(over=...)` or `split(by=...)` which would read somewhat more like English. I'm not sure how this would feel to non-native speakers, though.
   * I can't think of a good noun at the moment.
   * We could also eliminate the need to think about the name at all by making it a [positional-only parameter](https://www.python.org/dev/peps/pep-0570/), so it would always be `split("a")` or `split(["a", "b"])` or the like.

**Comment by mgxd** [comment]
> nr_proc is pretty unintuitive.

+1 on this, I'd be happy with either `n_procs` or `max_workers`

**Comment by djarecka** [comment]
>When splitting over multiple dimensions, the results are aggregated in a 1D list. I would have assumed that it would be ND, with the first axis matching the first variable split over, and so on.

ND list will be after using `combiner`. If we only split inputs, we create a flat list (with specific order)

>When combining, the order of the combiner does not affect the order of the results. I would assume that .combine(["a", "b"]) would produce something ordered like [(a1, b1), (a2, b1), (a1, b2), (a2, b2)] while .combine(["b", "a"]) would produce [(a1, b1), (a1, b2), (a2, b1), (a2, b2)]. (Or possibly the reverse; I haven't thought it through carefully enough to be sure.)

We at some point decided that the order of fields in a single combiner should not be relevant (see examples #126), but could be change at some point

> .combine("x").combine("n") seems like it should be equivalent to .combine(["x", "n"]). Probably an edge case, but I think making this (and multiple splits on a single task) doable should help clarify the semantics for users, as well as our own thinking. For instance, these should be equivalent (they don't actually run, but hopefully they're clear):

for now we don't allow for using `combine` twice for the same `Task` or `Workflow`.  If we decide to allow it, I'd say that `.combine("x").combine("n")` should give a different result than `.combine(["x", "n"])`, i.e. should be treated as two-steps combiner - `.combine("x")` would result in `[[x1n1, x2n1], [x1n2, x2n2]]` and `combine("n")`would additionally combine the two lists `[[[x1n1, x2n1], [x1n2, x2n2]]]` (notice the additional nested layer, would be more useful if we have more than two splitting fields).
And in that case your workflows `wf1` and `wf2` would be indeed equivalent (just likely giving a different output than you expected, `wf1` should run fine)
`wf3` introduces double `split` and this would require an additional discussion what does it mean, for now it's is also not allowed

>Combiners must be strings or lists, currently. It's not clear why this is a requirement. Any sequence seems acceptable to me.

what else would you use? that's the only types that make sense in the current implementation.

>Perhaps we can think of a more intuitive name from a user's perspective than splitter in Task().split(splitter=...). I think it's fine as an internal variable, so we don't need to do a massive search-and-replace, but splitter kind of grates on me.

I don't feel the same way, but if more people think that it should be changed, we can do it. But we don't have to use the name of the argument, you can use `split("a")` right now as well

**Comment by djarecka** [comment]
@satra @mgxd - these suggestions are more to the API than to the tutorial itself, so please comment on them as well if you have any opinions

**Comment by djarecka** [comment]
@effigies - I will keep this open, but I talked to @satra and we agreed that don't want to have double `split`/`combine` options, at least for now. We can come back to this later. 

For the `splitter` I don't have really opinion, but could change the tutorials to not use it for now.

**Comment by satra** [comment]
closed via #127
